src/shogun/multiclass/COVERTree.cpp

```diff
@@ -0,0 +1,137 @@
+#include <shogun/multiclass/KNNSolver.h>
+#include <shogun/lib/JLCoverTree.h>
+
+using namespace shogun;
+
+CCOVER_TREESolver::CCOVER_TREESolver(int32_t k, float64_t q, int32_t num_classes, int32_t min_label, SGVector<int32_t> train_labels)
+{
+	init();
+
+	m_k=k;
+	m_q=q;
+	m_num_classes=num_classes;
+	m_min_label=min_label;
+	m_train_labels=train_labels;
+}
+
+void CCOVER_TREESolver::init()
+{
+	m_k=3;
+	m_q=1.0;
+	m_num_classes=0;
+	m_min_label=0;
+	m_train_labels=0;
+}
+
+CMulticlassLabels* CCOVER_TREESolver::classify_objects(CDistance* distance, int32_t num_lab, int32_t* train_lab, float64_t* classes)
+{
+	CMulticlassLabels* output=new CMulticlassLabels(num_lab);
+
+	// m_q != 1.0 not supported with cover tree because the neighbors
+	// are not retrieved in increasing order of distance to the query
+	float64_t old_q = m_q;
+	if ( old_q != 1.0 )
+		SG_INFO("q != 1.0 not supported with cover tree, using q = 1\n")
+
+	// From the sets of features (lhs and rhs) stored in distance,
+	// build arrays of cover tree points
+	v_array< CJLCoverTreePoint > set_of_points  =
+		parse_points(distance, FC_LHS);
+	v_array< CJLCoverTreePoint > set_of_queries =
+		parse_points(distance, FC_RHS);
+
+	// Build the cover trees, one for the test vectors (rhs features)
+	// and another for the training vectors (lhs features)
+	CFeatures* r = distance->replace_rhs( distance->get_lhs() );
+	node< CJLCoverTreePoint > top = batch_create(set_of_points);
+	CFeatures* l = distance->replace_lhs(r);
+	distance->replace_rhs(r);
+	node< CJLCoverTreePoint > top_query = batch_create(set_of_queries);
+
+	// Get the k nearest neighbors to all the test vectors (batch method)
+	distance->replace_lhs(l);
+	v_array< v_array< CJLCoverTreePoint > > res;
+	k_nearest_neighbor(top, top_query, res, m_k);
+
+#ifdef DEBUG_KNN
+	SG_PRINT("\nJL Results:\n")
```

Comment:
use `SG_DEBUG` instead of `DEBUG_KNN`
make sure that before doing the nested for loop check the `msg_level`